run/markdown-preview/editor/src/main/resources/templates/index.html

```diff
@@ -0,0 +1,111 @@
+  </header>
+
+  <div role="progressbar" class="mdc-linear-progress mdc-linear-progress--indeterminate mdc-top-app-bar--fixed-adjust" aria-label="Markdown Rendering Progress Bar" aria-valuemin="0" aria-valuemax="1" aria-valuenow="0">
+    <div class="mdc-linear-progress__bar mdc-linear-progress__primary-bar">
+      <span class="mdc-linear-progress__bar-inner"></span>
+    </div>
+  </div>
+
+  <main class="mdc-layout-grid">
+    <div class="mdc-layout-grid__inner">
+      <div class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6">
+        <h2>Markdown Text</h2>
+        <section class="mdc-card mdc-card--outlined">
+          <div class="text-field-container">
+            <div class="mdc-text-field mdc-text-field--fullwidth md-text-field--no-label mdc-text-field--textarea mdc-ripple-upgraded">
+              <textarea id="editor" class="mdc-text-field__input" style="height: 36rem;" th:text="${Default}"></textarea>
+            </div>
+          </div>
+
+          <div class="mdc-card__actions mdc-card__actions--full-bleed">
+            <button class="editor-button mdc-button mdc-card__action mdc-card__action--button mdc-ripple-surface">
+              <span class="mdc-button__label">Preview Rendered Markdown</span>
+              <i class="material-icons" aria-hidden="true">arrow_forward</i>
+            </button>
+          </div>
+        </section>
+      </div>
+
+      <div class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6">
+        <h2>Rendered HTML</h2>
+        <section class="mdc-card mdc-card--outlined">
+          <div id="preview" style="height: 40rem; padding-left: 10px; padding-right: 10px">Tap "<strong>Preview Rendered Markdown</strong>" below the text entry to see rendered content.</div>
+        </section>
+      </div>
+    </div>
+  </main>
+
+  <script>
+    const preview = document.getElementById('preview');
+    const lp = new mdc.linearProgress.MDCLinearProgress(document.querySelector('.mdc-linear-progress'));
+    async function render(data = {}) {
+      const response = await fetch('/render', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify(data)
+      });
+
+      const text = await response.text();
+      if (!response.ok) {
+        console.log('error: Render Text: Received status code: ' + response.status);
+      }
+
+      return text;
+    }
+
+    function listener() {
+      lp.open();
+      render({data: document.getElementById('editor').value})
+      .then((result) => preview.innerHTML = result)
+      .catch((err) => {
+        console.log('Render Text: ' + err.message);
+        preview.innerHTML = '<h3><i aria-hidden="true" class="material-icons">error</i>Render Error</h3>\n<p>' + err.message + '</p>';
+      })
+      .finally(() => lp.close())
+    }
+
+    document.querySelector('.editor-button').addEventListener('click', listener);
+    window.addEventListener('load', listener);
+  </script>
```

Comment:
Do we have any CSRF or XSRF protection on this?

Reply:
The service uses the metadata server to add a token to make an authenticated request to the renderer service. I also sanitize the rendered HTML.